src/main/java/org/glite/authz/pep/pip/provider/policynamespip/PolicyNamesCache.java

```diff
@@ -0,0 +1,625 @@
+	    }
+	} catch (IOException e)	{
+	    // Try to close, this might throw a new IOException. We're throwing
+	    // one in any case.
+	    reader.close();
+	    throw new IOException("Reading from "+name+" failed: "+e.getMessage());
+	}
+
+	// Close reader
+	try {
+	    reader.close();
+	} catch (IOException e)	{
+	    throw new IOException("Closing "+name+" failed: "+e.getMessage());
+	}
+
+	// All done
+	return value;
+    }
+
+    /**
+     * Parses out the different components of a subjectdn value as obtained by
+     * {@link #getSubjectDNvalue(Path)} and puts the results into the given entry.
+     * When a "file:" entry is found, it will add the correct path to the right
+     * list of dependencies.
+     * @param entry {@link Entry} to fill
+     * @param value value of the subjectdn key
+     * @throws ParseException on syntax errors in the value
+     * @throws IOException on read errors with one of the info files
+     */
+    private void parseSubjectDNvalue(Entry entry, String value)
+	throws ParseException, IOException
+    {
+	// Now parse the value part
+	int pos=0, pos2;
+	while (true)	{
+	    // Check what we have here
+	    if (value.charAt(pos)=='"')	{
+		// Found CA subject DN: look for end quote (pos==quote_1)
+		pos2=value.indexOf('"', pos+1);
+		if ( pos2 < 0 )
+		    throw new ParseException(entry.name+": Missing end-quote", pos+1);
+		// Add url-decoded value to the subDN list
+		entry.localSubDNs.add(URLDecoder.decode(value.substring(pos+1,pos2),"UTF-8"));
+		// skip over end-quote
+		pos2++;
+	    } else if (value.substring(pos).startsWith("file:"))    {
+		// Found file: look for end of filename
+		for (pos2=pos+5;
+		     pos2<value.length() && value.charAt(pos2)!=',' && value.charAt(pos2)!=' ' && value.charAt(pos2)!='\t';
+		     pos2++);
+
+		// Add value to the right dependency lists
+		entry.addDependency(value.substring(pos+5,pos2));
+	    } else {
+		// Found unknown character at the start
+		throw new ParseException(entry.name+": "+SUBJECT_KEY+" value invalid: "+value, pos);
+	    }
+
+	    // Skip all trailing white-space and commas
+	    boolean foundComma=false;
+	    for (pos=pos2; pos<value.length(); pos++)    {
+		// Keep track of whether we found a comma: need at least one
+		if (value.charAt(pos)==',') {
+		    foundComma=true;
+		    continue;
+		}
+		if (value.charAt(pos)!=' ' && value.charAt(pos)!='\t')
+		    break;
+	    }
+
+	    // Did we hit the end-of-line or find a comment char?
+	    if (pos==value.length() || value.charAt(pos)=='#')
+		break;
+
+	    // Not yet at end, did we see a comma?
+	    if (!foundComma)
+		throw new ParseException(entry.name+": Missing comma delimiter before new entry", pos);
+	}
+    }
+
+
+    ////////////////////////////////////////////////////////////////////////
+    // Private class
+    ////////////////////////////////////////////////////////////////////////
+
+    /** Internal representation of info file entries */
+    private class Entry	{
```

Comment:
I don't think so: it is also using the trustDir, infoEntries and extInfoEntries from its parent and they should be instance variables, not class variables.